### ISSUE TITLE
[ES|QL] Support for shadowing warning

### ISSUE THREAD

**Issue by dej611** [elastic/elasticsearch#100331]
### Description

It would be nice to have a warning in case the user is overriding a field (in particularly changing its type).

Assume an `index` with a `bytes: double` field.

```
from index | eval bytes = "blah"
```

It would be nice in this scenario to warn the user about the existence of the `bytes` field and its overriding, or at notify about the type change.
Even a message like would be nice: `Column [bytes] of type "number" has been overwritten with new type: "string"`

I've put a PoC in Kibana side and it is quite useful when working with an index a user is not familiar with, so perhaps it can be useful to consider at validation level too:
<img width="825" alt="Screenshot 2023-10-05 at 11 47 46" src="https://github.com/elastic/elasticsearch/assets/924948/a22df523-42c2-4727-b71c-2d4f336a3f3e">

**Comment by nik9000** [comment]
Interesting! I think it's going to be pretty normal to want to convert a field's type with something like `EVAL n = TO_STRING(n)`. So the scenario might be a bit more complex than what you say. But it does sound like useful feedback.

**Comment by dej611** [comment]
> I think it's going to be pretty normal to want to convert a field's type with something like `EVAL n = TO_STRING(n)`

I completely agree. Having it as a warning will let the user know without preventing him from doing so. Also, at Kibana level we can decide whether or not show such warning (or even have a more fine control over it).
